reassembly/tcpassembly.go

```diff
@@ -0,0 +1,1319 @@
+	Seq() Sequence
+	IsPacket() bool
+}
+
+// Implements a ScatterGather
+type reassemblyObject struct {
+	all       []byteContainer
+	Skip      int
+	Direction TCPFlowDirection
+	saved     int
+	toKeep    int
+	// stats
+	queuedBytes    int
+	queuedPackets  int
+	overlapBytes   int
+	overlapPackets int
+}
+
+func (rl *reassemblyObject) Lengths() (int, int) {
+	l := 0
+	for _, r := range rl.all {
+		l += r.Length()
+	}
+	return l, rl.saved
+}
+
+func (rl *reassemblyObject) Fetch(l int) []byte {
+	if len(rl.all) == 1 && l == rl.all[0].Length() {
+		return rl.all[0].Bytes()
+	}
+	bytes := []byte{}
+	for _, bc := range rl.all {
+		bytes = append(bytes, bc.Bytes()...)
+	}
+	return bytes[:l]
+}
+
+func (rl *reassemblyObject) KeepFrom(offset int) {
+	rl.toKeep = offset
+}
+
+func (rl *reassemblyObject) CaptureInfo(offset int) gopacket.CaptureInfo {
+	current := 0
+	for _, r := range rl.all {
+		if current >= offset {
+			return r.CaptureInfo()
+		}
+		current += r.Length()
+	}
+	// Invalid offset
+	return gopacket.CaptureInfo{}
+}
+
+func (rl *reassemblyObject) Info() (TCPFlowDirection, bool, bool, int) {
+	return rl.Direction, rl.all[0].Start(), rl.all[len(rl.all)-1].End(), rl.Skip
+}
+
+func (rl *reassemblyObject) Stats() TCPAssemblyStats {
+	packets := int(0)
+	for _, r := range rl.all {
+		if r.IsPacket() {
+			packets++
+		}
+	}
+	return TCPAssemblyStats{
+		Chunks:         len(rl.all),
+		Packets:        packets,
+		QueuedBytes:    rl.queuedBytes,
+		QueuedPackets:  rl.queuedPackets,
+		OverlapBytes:   rl.overlapBytes,
+		OverlapPackets: rl.overlapPackets,
+	}
+}
+
+const pageBytes = 1900
+
+// TCPFlowDirection distinguish the two half-connections directions.
+//
+// TCPDirClientToServer is assigned to half-connection for the first received
+// packet, hence might be wrong if packets are not received in order.
+// It's up to the caller (e.g. in Accept()) to decide if the direction should
+// be interpretted differently.
+type TCPFlowDirection bool
+
+// Value are not really useful
+const (
+	TCPDirClientToServer TCPFlowDirection = false
+	TCPDirServerToClient TCPFlowDirection = true
+)
+
+func (dir TCPFlowDirection) String() string {
+	switch dir {
+	case TCPDirClientToServer:
+		return "client->server"
+	case TCPDirServerToClient:
+		return "server->client"
+	}
+	return ""
+}
+
+// Reverse returns the reversed direction
+func (dir TCPFlowDirection) Reverse() TCPFlowDirection {
+	return !dir
+}
+
+/* page: implements a byteContainer */
+
+// page is used to store TCP data we're not ready for yet (out-of-order
+// packets).  Unused pages are stored in and returned from a pageCache, which
+// avoids memory allocation.  Used pages are stored in a doubly-linked list in
+// a connection.
+type page struct {
+	bytes []byte
+	seq   Sequence
+	//index      int
+	prev, next *page
+	buf        [pageBytes]byte
+	ac         AssemblerContext
+	seen       time.Time
+	start, end bool
+	isPacket   bool // only the first page of a packet has this being set
```

Comment:
It is just used to count the number of TCP packets for a given chunk given to the caller, changing to a pointer seems a bit overkill.
Using ac as boolean is doable: Done.